### PR TITLE
[Do Not Review or Merge] Support Custom Architectures

### PR DESCRIPTION
**Pull request by jamesmcclain**
## Overview

Brief description of what this PR does, and why it is needed.

### Checklist

- [ ] Updated `docs/changelog.rst`
- [ ] Added `needs-backport` label if PR is bug fix that applies to previous minor release
- [ ] Ran scripts/format_code and committed any changes
- [ ] Documentation updated if needed
- [ ] PR has a name that won't get you publicly shamed for vagueness

### Notes

Optional. Ancillary topics, caveats, alternative strategies that didn't work out, anything else.

## Testing Instructions

```python
# tiny_spacenet.py

from os.path import join

from rastervision2.core.rv_pipeline import *
from rastervision2.core.backend import *
from rastervision2.core.data import *
from rastervision2.research_backend import *
# from rastervision2.pytorch_learner import *


def get_config(runner):
    root_uri = '/opt/data/output/'
    base_uri = ('https://s3.amazonaws.com/azavea-research-public-data/'
                'raster-vision/examples/spacenet')
    train_image_uri = '{}/RGB-PanSharpen_AOI_2_Vegas_img205.tif'.format(
        base_uri)
    train_label_uri = '{}/buildings_AOI_2_Vegas_img205.geojson'.format(
        base_uri)
    val_image_uri = '{}/RGB-PanSharpen_AOI_2_Vegas_img25.tif'.format(base_uri)
    val_label_uri = '{}/buildings_AOI_2_Vegas_img25.geojson'.format(base_uri)
    channel_order = [0, 1, 2]
    class_config = ClassConfig(
        names=['building', 'background'], colors=['red', 'black'])

    def make_scene(scene_id, image_uri, label_uri):
        """
        - StatsTransformer is used to convert uint16 values to uint8.
        - The GeoJSON does not have a class_id property for each geom,
        so it is inferred as 0 (ie. building) because the default_class_id
        is set to 0.
        - The labels are in the form of GeoJSON which needs to be rasterized
        to use as label for semantic segmentation, so we use a RasterizedSource.
        - The rasterizer set the background (as opposed to foreground) pixels
        to 1 because background_class_id is set to 1.
        """
        raster_source = RasterioSourceConfig(
            uris=[image_uri],
            channel_order=channel_order,
            transformers=[StatsTransformerConfig()])
        label_source = SemanticSegmentationLabelSourceConfig(
            raster_source=RasterizedSourceConfig(
                vector_source=GeoJSONVectorSourceConfig(
                    uri=label_uri, default_class_id=0, ignore_crs_field=True),
                rasterizer_config=RasterizerConfig(background_class_id=1)))
        return SceneConfig(
            id=scene_id,
            raster_source=raster_source,
            label_source=label_source)

    dataset = DatasetConfig(
        class_config=class_config,
        train_scenes=[
            make_scene('scene_205', train_image_uri, train_label_uri)
        ],
        validation_scenes=[
            make_scene('scene_25', val_image_uri, val_label_uri)
        ])

    # Use the PyTorch backend for the SemanticSegmentation pipeline.
    chip_sz = 300
    backend = ResearchArchitectureConfig(
        architecture='https://raw.githubusercontent.com/geotrellis/deeplab-nlcd/b203f716d381cf4e908b298c3f55e14ceba076b3/python/architectures/deeplab-resnet18.py',
        pretrained = True,
        bands = 3,
        resolution_divisor = 8,
        solver=SolverConfig(lr=1e-4, num_epochs=1, batch_sz=2))
    chip_options = SemanticSegmentationChipOptions(
        window_method=SemanticSegmentationWindowMethod.random_sample,
        chips_per_scene=10)

    return SemanticSegmentationConfig(
        root_uri=root_uri,
        dataset=dataset,
        backend=backend,
        train_chip_sz=chip_sz,
        predict_chip_sz=chip_sz,
        chip_options=chip_options)
```

1. `docker run -it --rm -v $HOME/.aws:/root/.aws:ro -v $(pwd):/opt/src -v /tmp/rastervision2:/tmp/rastervision2:ro quay.io/azavea/raster-vision:pytorch-latest bash`
2. `rastervision2 --profile local run inprocess /tmp/rastervision2/pipeline/tiny_spacenet.py`

### CODE REVIEW

**Comment by jamesmcclain** [comment]
Moving this elsewhere.

**Comment by jamesmcclain** [comment]
I am reopening this in order to make sure that this code does not get lost.  Do not review or merge.

**Comment by jamesmcclain** [comment]
This has been superseded by #981
